teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,502 @@
+{"name": "us-west4-maven.pkg.dev/test-project/test-repo/com/google/guava/guava/31.0/guava-31.0.pom", "digest": {"sha256": "def"}}
+{"name": "us-west4-maven.pkg.dev/test-project/test-repo/com/google/guava/guava/31.0/guava-31.0-sources.jar", "digest": {"sha256": "xyz"}}]
+...
+materials: [{
+  // The git repo that contains the build.yaml referenced above.
+  "uri": "http://myp4web:8080/depot/main/atlas/",
+  // The resolved git commit hash reflecting the version of the repo used
+  // for this build.
+  "digest": {"sha1": "abc..."}
+}]
+```
+### Requirements
+
+- Able to structurally retrieve provenance info from TaskRun and produces currently supported provenance formats once a TaskRun is finished
+- Based on SLSA L2 requirements, the provenance must provide the location of source code in version control and the provenance MUST identify the output artifact via at least one cryptographic hash.
+- If a TaskRun is within a PipelineRun, the TaskRun’s provenance can be produced between the time that when one TaskRun is finished
+
+
+## Proposal
+
+Currently we are using type-hinting to retrieve needed provenance info from TaskRun’s results. The proposal aims to leverage the new features of structured results described in [TEP 0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) and [TEP 0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md) and support better structured provenance retrieval for inputs and artifacts to make the current type-hinting easier. This proposal is focused on the use cases for generation in-toto attestations, as other types of attestations don’t need the TaskRun results.
+
+The workflow for the targeted results in Chains is like the following:
+```
+(TaskRun results) —-> (Scanned by Chains controller, targets are collected) —> (Targets are parsed and validated and matched against with defined result and param structures)  —-> (Chains controller sign the signable targets and put into intoto attestations)
+```
+Tekton Chains will follow [TEP 0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) and [TEP 0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md) and support the sets of json schemas and described in the documents. In this proposal, the expected formats of results are also defined.
+
+As mentioned above, the signable targets can be split into inputs (sources) and outputs (artifacts). It's important that Chains are able to determine which signable targets are inputs, and which are outputs. The signable target result names are all captured in either of the results, `ARTIFACT_INPUTS` or `ARTIFACT_OUTPUTS`, which will be two arrays storing the result names that should be captured by Tekton Chains to format in-toto provenances. `ARTIFACT_INPUTS` will be captured in `subjects` and `ARTIFACT_OUTPUTS` will be captured in `materials`.
```

Comment:
This line says 
> `ARTIFACT_INPUTS` will be captured in `subjects` and `ARTIFACT_OUTPUTS` will be captured in `materials`.

But the example above is inconsistent. In the example, items from `ARTIFACT_INPUTS` are stored into `materials`, whereas items from `ARTIFACT_OUTPUTS` are stored into `subjects`. Which one is intended?

Also can you please help me understand a bit more about the differences between `ARTIFACT_INPUTS` and `ARTIFACT_OUTPUTS` given that they are all defined in `results`? Thanks

Reply:
Good catch!

Reply:
In terms of differences of `ARTIFACT_INPUTS` and `ARTIFACT_OUTPUTS`, they mean that whether an artifact A is used as source or materials in a pipeline in order to generate another artifact B. In this case, A is `ARTIFACT_INPUTS` and B is `ARTIFACT_OUTPUTS`